File: everyDay/1138.字母板上的路径.go
/*
 * @lc app=leetcode.cn id=1138 lang=golang
 *
 * [1138] 字母板上的路径
 */

// @lc code=start
func alphabetBoardPath(target string) string {
	a := []byte{}
	x, y := 0, 0
	for _, i := range target {
		tx := int(i - 'a') % 5
		ty := int(i - 'a') / 5
		if tx == x && ty == y {
			a = append(a, '!')
		} else {
			if (ty - y < 0) {
				for j := 0; j < y - ty; j++ {
					a = append(a, 'U')
				}
			}
			if (tx - x < 0) {
				for j := 0; j < x - tx; j++ {
					a = append(a, 'L')
				}
			}
			if (tx - x >= 0) {
				for j := 0; j < tx - x; j++ {
					a = append(a, 'R')
				}
			}
			if (ty - y >= 0) {
				for j := 0; j < ty - y; j++ {
					a = append(a, 'D')
				}
			}
			a = append(a, '!')
		}
		x , y = tx, ty
	}
	return string(a)
}
// @lc code=end

File: everyDay/1138.字母板上的路径.cpp
/*
 * @lc app=leetcode.cn id=1138 lang=cpp
 *
 * [1138] 字母板上的路径
 */

// @lc code=start
class Solution {
public:
    string alphabetBoardPath(string target) {
        /*
        v / 5 = y
        v % 5 = x
        */
        string ans = "";
        int p = 0;
        for (auto it : target) {
            
            int t = it - 'a';
            if (p == t) {
                ans += '!';
            } else {
                int ty = t / 5, tx = t % 5;
                int py = p / 5, px = p % 5;
                if (ty - py > 0) {
                    if (tx - px > 0) {
                        for (int j = 0; j < tx - px; j++) {
                            ans += 'R';
                        }
                    } else {
                        for (int j = 0; j < abs(tx - px); j++) {
                            ans += 'L';
                        }
                    }
                    for (int j = 0; j < ty - py; j++) {
                        ans += 'D';
                    }
                } else {
                    for (int j = 0; j < abs(ty - py); j++) {
                        ans += 'U';
                    }
                    if (tx - px > 0) {
                        for (int j = 0; j < tx - px; j++) {
                            ans += 'R';
                        }
                    } else {
                        for (int j = 0; j < abs(tx - px); j++) {
                            ans += 'L';
                        }
                    }
                } 
                ans += '!';
            }
            p = t;
        }
        return ans;
    }
};
// @lc code=end

